src/arithmetic/fields.rs

```diff
@@ -85,6 +103,119 @@ pub trait FieldExt:
         }
     }
 
+    /// Computes:
+    ///
+    /// * (true,  sqrt(num/div)),                 if num and div are nonzero and num/div is a square in the field;
+    /// * (true,  0),                             if num is zero;
+    /// * (false, 0),                             if num is nonzero and div is zero;
+    /// * (false, sqrt(ROOT_OF_UNITY * num/div)), if num and div are nonzero and num/div is a nonsquare in the field;
+    ///
+    /// where ROOT_OF_UNITY is a generator of the order 2^n subgroup (and therefore a nonsquare).
+    ///
+    /// The choice of root from sqrt is unspecified.
+    fn sqrt_ratio(num: &Self, div: &Self) -> (Choice, Self) {
+        // Based on:
+        // * [Sarkar2020](https://eprint.iacr.org/2020/1407)
+        // * [BDLSY2012](https://cr.yp.to/papers.html#ed25519)
+        //
+        // We need to calculate uv and v, where v = u^((m-1)/2), u = num/div, and p-1 = T * 2^S.
+        // We can rewrite as follows:
+        //
+        //      v = (num/div)^((T-1)/2)
+        //        = num^((T-1)/2) * div^(p-1 - (T-1)/2)    [Fermat's Little Theorem]
+        //        =       "       * div^(T * 2^S - (T-1)/2)
+        //        =       "       * div^((2^(S+1) - 1)*(T-1)/2 + 2^S)
+        //        = (num * div^(2^(S+1) - 1))^((T-1)/2) * div^(2^S)
+        //
+        // Let  w = (num * div^(2^(S+1) - 1))^((T-1)/2) * div^(2^S - 1).
+        // Then v = w * div, and uv = num * v / div = num * w.
+        //
+        // We calculate:
+        //
+        //      s = div^(2^S - 1) using an addition chain
+        //      t = div^(2^(S+1) - 1) = s^2 * div
+        //      w = (num * t)^((T-1)/2) * s using another addition chain
+        //
+        // then u and uv as above. The addition chains are given in
+        // https://github.com/zcash/pasta/blob/master/addchain_sqrt.py .
+        // The overall cost of this part is similar to a single full-width exponentiation,
+        // regardless of S.
+
+        let sqr = |x: Self, i: u32| (0..i).fold(x, |x, _| x.square());
+
+        // s = div^(2^S - 1)
+        let s = (0..5).fold(*div, |d: Self, i| sqr(d, 1 << i) * d);
```

Comment:
```suggestion
        // s = div^(2^S - 1)
        let s = div.pow_by_2toS_minus1();
```
But @str4d suggests not doing this for now, because Pallas and Vesta have S = 32.

Reply:
Yeah, I feel that these kinds of APIs would not belong in `ff::Field`, and thus should be discouraged in `FieldExt` (the latter is an internal proving ground, but ideally we would not need it). And that we probably shouldn't actually be providing a default implementation of `sqrt_div`, since it necessarily requires exposing implementation-specific details to downstream users.

Instead (as discussed in PR gardening today) I think it would make more sense to make the `SqrtTable` struct be `Sakhar2020Sqrt` or something, have its init function take in e.g. a perfect hashing function for the given field, and then have that struct provide this implementation (taking in any necessary addition chains). Then a field implementor who wants to use this algorithm can just plug the pieces together to implement `sqrt_div`.

Reply:
Done in 8350773aeaf6dd0012c0afdd9cd94b1075ddfc27 (I haven't included the addition chains yet, but they're easy to do from here).